### PR TITLE
feat: Use pgbouncer for prisma client

### PR DESCRIPTION
**Pull request by istarkov**
## Description

Add pgbouncer
- [x] - Checked that on supabase connections goes through pgbouncer

closes https://github.com/webstudio-is/webstudio-builder/issues/1878

Tested using following k6
```js
import encoding from 'k6/encoding'
import http from 'k6/http'
import { check } from 'k6'

export const options = {
  vus: 100,
  duration: '20s',
}

export default function () {
  const res = http.get("https://webstudio-builder-git-bgbouncer-webstudio-is.vercel.app/builder/cddc1d44-af37-4cb6-a430-d300cf6f932d?authToken=1cdc6026-dd5b-4624-b89b-9bd45e9bcc3d&mode=preview", {
  "headers": {
    "accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7",
    "accept-language": "en-US,en;q=0.9",
    "cache-control": "max-age=0",
    "sec-ch-prefers-color-scheme": "dark",
    "sec-ch-ua": "\"Chromium\";v=\"116\", \"Not)A;Brand\";v=\"24\", \"Google Chrome\";v=\"116\"",
    "sec-ch-ua-mobile": "?0",
    "sec-ch-ua-platform": "\"macOS\"",
    "sec-fetch-dest": "document",
    "sec-fetch-mode": "navigate",
    "sec-fetch-site": "same-origin",
    "sec-fetch-user": "?1",
    "upgrade-insecure-requests": "1"
  },
});

  // Verify response (checking the echoed data from the httpbin.org basic auth test API endpoint)
  check(res, {
    'status is 200': (r) => r.status === 200,
    'has data': (r) => r.body.includes("rootInstanceId"),    
  })
}
```

Increased pg_bouncer max_client_conn from 100 to 400 based on results

## Steps for reproduction

1. click button
2. expect xyz

## Code Review

- [ ] hi @kof, I need you to do
  - conceptual review (architecture, feature-correctness)
  - detailed review (read every line)
  - test it on preview

## Before requesting a review

- [ ] made a self-review
- [ ] added inline comments where things may be not obvious (the "why", not "what")

## Before merging

- [ ] tested locally and on preview environment (preview dev login: 5de6)
- [ ] updated [test cases](https://github.com/webstudio-is/webstudio-builder/blob/main/apps/builder/docs/test-cases.md) document
- [ ] added tests
- [ ] if any new env variables are added, added them to `.env.example` and the `builder/env-check.js` if mandatory

### CODE REVIEW

File: packages/prisma-client/src/prisma.ts [hunk 1]
```diff
@@ -16,13 +16,37 @@ declare global {
 
 const logPrisma = process.env.NODE_ENV === "production";
 
+const getPgBouncerUrl = () => {
+  // The URL constructor does not support the "postgresql" protocol, so we replace it with "https"
+  // to ensure that the connection is made over HTTPS instead of plain text.
+  const databaseUrl = new URL(
+    process.env.DATABASE_URL!.replace("postgresql://", "https://")
```
Comment (kof) [inline packages/prisma-client/src/prisma.ts:23]: I would throw instead of silencing types

File: packages/prisma-client/src/prisma.ts [hunk 2]
```diff
@@ -16,13 +16,37 @@ declare global {
...
+  );
+
+  if (databaseUrl.hostname.endsWith("supabase.co")) {
```
Comment (kof) [inline packages/prisma-client/src/prisma.ts:26]: I don't like this hardcoding at all ...

Comment (kof) [reply packages/prisma-client/src/prisma.ts:26]: we shoudln't be bound to supabase in any way, in fact its likely that we are going to switch

File: packages/prisma-client/src/prisma.ts (hunk 1 above)
Comment (istarkov) [reply packages/prisma-client/src/prisma.ts:23]: We already check it at env check.
I would not init prisma client here at all. Must be rewritten to init at main app

File: packages/prisma-client/src/prisma.ts (hunk 2 above)
Comment (kof) [reply packages/prisma-client/src/prisma.ts:26]: this should be completely coming from env

File: packages/prisma-client/src/prisma.ts (hunk 1 above)
Comment (kof) [reply packages/prisma-client/src/prisma.ts:23]: sure, but throwing it makes it clear

File: packages/prisma-client/src/prisma.ts (hunk 2 above)
Comment (istarkov) [reply packages/prisma-client/src/prisma.ts:26]: I as devops don't like to have 100s of env variables.

Comment (istarkov) [reply packages/prisma-client/src/prisma.ts:26]: And here we don't bound to supabase, we just use pgbouncer if supabase database is used.

**Review by TrySound — APPROVED** [review]

**Comment by kof** [comment]
Discussed
- [ ] add PGBOUNCER=true env
- [ ] remove https protocol replacement workarounds, parser works in node correctly, its a bug in chrome  https://bugs.chromium.org/p/chromium/issues/detail?id=869291
- [ ] remove the supabase check

**Comment by TrySound** [comment]
> remove https protocol replacement workarounds, parser works in node correctly, its a bug in chrome

Aha!

**Pull request merged**